pkg/cmd/psa/psa.go

```diff
@@ -0,0 +1,302 @@
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	psapi "k8s.io/pod-security-admission/api"
+	utilpointer "k8s.io/utils/pointer"
+
+	"github.com/spf13/cobra"
+)
+
+// PSAOptions contains all the options and configsi for running the PSA command.
+type PSAOptions struct {
+	quiet     bool
+	level     string
+	namespace string
+
+	configFlags genericclioptions.ConfigFlags
+	client      *kubernetes.Clientset
+	warnings    *warningsHandler
+}
+
+var (
+	psaExample = `
+	# Check if all cluster namespaces can be upgraded to the 'restricted' security level.
+	%[1]s psa-check --level restricted
+
+	# Check if a specific namespace, 'my-namespace', can be upgraded to the 'restricted' security level.
+	%[1]s psa-check --level restricted --namespace my-namespace
+`
+
+	empty       = struct{}{}
+	validLevels = map[string]struct{}{
+		string(psapi.LevelPrivileged): empty,
+		string(psapi.LevelBaseline):   empty,
+		string(psapi.LevelRestricted): empty,
+	}
+
+	podControllers = map[string]struct{}{
+		"Deployment":  empty,
+		"DaemonSet":   empty,
+		"StatefulSet": empty,
+		"CronJob":     empty,
+		"Job":         empty,
+	}
+)
+
+// NewCmdPSA creates a new cobra.Command instance that enables checking
+// namespaces for their compatibility with a specified PodSecurity level.
+func NewCmdPSA(parentName string, streams genericclioptions.IOStreams) *cobra.Command {
+	o := PSAOptions{
+		configFlags: genericclioptions.ConfigFlags{
+			Namespace:  utilpointer.StringPtr(""),
+			KubeConfig: utilpointer.StringPtr(""),
+		},
+	}
+
+	cmd := cobra.Command{
+		Use:   "psa-check",
+		Short: "Verify namespace workloads match the namespace pod security profile",
+
+		Example: fmt.Sprintf(psaExample, parentName),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Validate(); err != nil {
+				return fmt.Errorf("validation failed: %w", err)
+			}
+			if err := o.Complete(); err != nil {
+				return fmt.Errorf("completion failed: %w", err)
+			}
+
+			return o.Run()
+		},
+	}
+
+	fs := cmd.Flags()
+	o.configFlags.AddFlags(fs)
+	fs.StringVar(&o.level, "level", "restricted", "The PodSecurity level to check against.")
+	fs.BoolVar(&o.quiet, "quiet", false, "Do not return non-zero exit code on violations.")
+
+	return &cmd
+}
+
+// Validate ensures that all required arguments and flag values are set properly.
+func (o *PSAOptions) Validate() error {
+	if _, ok := validLevels[o.level]; !ok {
+		return fmt.Errorf("invalid level %q", o.level)
+	}
+
+	return nil
+}
+
+// Complete sets all information required for processing the command.
+func (o *PSAOptions) Complete() error {
+	config := o.configFlags.ToRawKubeConfigLoader()
+	restConfig, err := config.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create rest config: %w", err)
+	}
+
+	ns, set, err := config.Namespace()
+	if err != nil {
+		return fmt.Errorf("failed to get namespace: %w", err)
+	}
+	if set {
```

Comment:
In the current implementation, if the namespace flag is not set, I will use all namespaces. However, as @deads2k pointed out, it would be more appropriate to use the default namespace when the flag is not set and only search for all namespaces if someone actively sets the `--all-namespaces` flag.

I'm considering two approaches for handling the `--all-namespaces` flag: either [manually handling it](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/label/label.go#L158) or using the [ResourceBuilderFlags](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/builder_flags.go#L126-L128). I'm open to suggestions and feedback on which approach would be more suitable.

While using the `ResourceBuilderFlags` would align more with the builder pattern used in `kubectl` / `oc`, it would require a larger change. Given that `kubectl` doesn't strictly follow this pattern, I would suggest handling the flag manually for now.

Reply:
I think, you can continue managing your own all-namespace flag as in here https://github.com/kubernetes/kubernetes/blob/7da3893003eb2e5007a45f38b1276d554125c5b8/staging/src/k8s.io/kubectl/pkg/cmd/events/events.go#L144, since you don't use builder.
If user does not pass `-A` or `-n`, it should use default namespace;
If user passes `-n` explicitly, it should use it;
If user passes `-A`, ns should be empty and all namespaces should be used.